week4_assignment/login-signup-project/src/pages/Signup.jsx

```diff
@@ -0,0 +1,219 @@
+import styled from 'styled-components';
+import { useNavigate } from 'react-router-dom';
+import { useState, useRef, useEffect } from 'react';
+
+function SignUp() {
+  const [id, setId] = useState('');
+  const [pw, setPw] = useState('');
+  const [nickname, setNickname] = useState('');
+  const [phone, setPhone] = useState('');
+  const navigate = useNavigate();
+
+  const idInputRef = useRef(null);
+  const pwInputRef = useRef(null);
+  const nicknameInputRef = useRef(null);
+  const phoneInputRef = useRef(null);
+
+  //전화번호 양식을 정규표현식으로 자동입력
+  const handlePhoneInput = (event) => {
+    let value = event.target.value;
+    const numbers = value.replace(/[^\d]/g, '');
+    let formatted = '';
+    if (numbers.length > 0) {
+      formatted = numbers.substring(0, 3);
+      if (numbers.length >= 4) {
+        formatted += '-' + numbers.substring(3, 7);
+      }
+      if (numbers.length >= 8) {
+        formatted += '-' + numbers.substring(7, 11);
+      }
+      setPhone(formatted);
+    } else {
+      setPhone('');
+    }
+  };
+
+  const validateForm = () => {
+    const idPattern = /^[a-zA-Z0-9_]+$/;
+    const pwPattern = /^(?=.*[A-Za-z])(?=.*[0-9])(?=.*[!@#$%^&*])[a-zA-Z0-9!@#$%^&*]{8,}$/;
+    const phonePattern = /^010-\d{4}-\d{4}$/;
+
+    let isValid = true;
+
+    if (!id) {
+      idInputRef.current.style.borderColor = 'red';
+      idInputRef.current.focus();
+      isValid = false;
+    } else {
+      idInputRef.current.style.borderColor = 'black';
+    }
+
+    if (!pw) {
+      pwInputRef.current.style.borderColor = 'red';
+      if (isValid) { 
+        pwInputRef.current.focus();
+      }
+      isValid = false;
+    } else {
+      pwInputRef.current.style.borderColor = 'black'; 
+    }
+
+    if (!nickname) {
+      nicknameInputRef.current.style.borderColor = 'red';
+      if (isValid) { 
+        nicknameInputRef.current.focus();
+      }
+      isValid = false;
+    } else {
+      nicknameInputRef.current.style.borderColor = 'black'; 
+    }
+
+    if (!phone) {
+      phoneInputRef.current.style.borderColor = 'red';
+      if (isValid) { 
+        phoneInputRef.current.focus();
+      }
+      isValid = false;
+    } else {
+      phoneInputRef.current.style.borderColor = 'black';
+    }
+
+    if (isValid && (!idPattern.test(id) || !pwPattern.test(pw) || !phonePattern.test(phone))) {
+      isValid = false;
+    }
+
+    return isValid;
+};
```

Comment:
이 로직 이렇게 추상화시켜서 유지보수 및 가독성을 높일 수 있을 것 같아요!!
```suggestion
const validateForm = () => {
  const idPattern = /^[a-zA-Z0-9_]+$/;
  const pwPattern = /^(?=.*[A-Za-z])(?=.*[0-9])(?=.*[!@#$%^&*])[a-zA-Z0-9!@#$%^&*]{8,}$/;
  const phonePattern = /^010-\d{4}-\d{4}$/;

  const inputs = [
    { ref: idInputRef, value: id, pattern: idPattern },
    { ref: pwInputRef, value: pw, pattern: pwPattern },
    { ref: nicknameInputRef, value: nickname },
    { ref: phoneInputRef, value: phone, pattern: phonePattern }
  ];

  let isValid = true;
  let focusSet = false;

  inputs.forEach(input => {
    const { ref, value, pattern } = input;

    if (!value) {
      ref.current.style.borderColor = 'red';
      if (!focusSet) {
        ref.current.focus();
        focusSet = true;
      }
      isValid = false;
    } else {
      ref.current.style.borderColor = 'black';
    }

    if (pattern && !pattern.test(value)) {
      isValid = false;
    }
  });

  return isValid;
};

```

Reply:
우와 확실히 깔끔하네요.... 👍